src/scatter.h

```diff
@@ -33,222 +35,335 @@ template <class T = float> class FormFactorSphere {
+            }
+            T intensity_corr = 0;
+            if (r_cutoff < r_cutoff_infty && volume > 0) {
+                intensity_corr = 4 * pc::pi * N / (volume * std::pow(q, 3)) *
+                         (q * r_cutoff * std::cos(q * r_cutoff) - std::sin(q * r_cutoff));
+            }
+            sampling[m] += weight;
+            intensity[m] += ((2 * intensity_sum[m] + intensity_self_sum) / N + intensity_corr) * weight;
         }
     }
 
     /**
-     * @brief Sample between all groups
-     *
-     * Instead of looping over all particles, this will ignore all internal
-     * group distances.
-     *
-     * @param p Particle vector
-     * @param groupList Vector of group pointers - i.e. as returned from `Space::groupList()`.
-     * @note Particle form factors are always set to unity, i.e. F(q) is ignored.
-     *
-     * @warning Untested
+     * @return a tuple of min, max, and step parameters of a q-mash
      */
-    template <class Tpvec, class Tg> void sampleg2g(const Tpvec &p, Tg groupList) {
-        assert(qmin > 0 && qmax > 0 && dq > 0 && "q range invalid.");
-        sampleg2g(p, qmin, qmax, dq, groupList);
+    std::tuple<T, T, T> getQMeshParameters() {
+        return std::make_tuple(q_mesh_min, q_mesh_max, q_mesh_step);
     }
 
-    template <class Tpvec, class Tg> void sampleg2g(const Tpvec &p, T qmin, T qmax, T dq, Tg groupList) {
-        std::vector<T> _I(int((qmax - qmin) / dq + 0.5));
-        // loop over all pairs of groups, then over particles
-        for (int k = 0; k < (int)groupList.size() - 1; k++)
-            for (int l = k + 1; l < (int)groupList.size(); l++)
-                for (auto i : *groupList.at(k))
-                    for (auto j : *groupList.at(l)) {
-                        T r = geo.vdist(p[i].pos, p[j].pos).norm();
-                        int cnt = 0;
-                        for (T q = qmin; q <= qmax; q += dq)
-                            _I.at(cnt++) += sin(q * r) / (q * r);
-                    }
-        int cnt = 0, N = p.size();
-        for (T q = qmin; q <= qmax; q += dq)
-            I[q] += 2. * _I.at(cnt++) / N + 1; // add to average I(q)
+    /**
+     * @return a map containing q (key) and average intensity (value)
+     */
+    std::map<double, double> getIntensity() {
+        std::map<double, double> averaged_intensity;
+        for (int m = 0; m < intensity.size(); ++m) {
+            const T average = intensity[m] / (sampling[m] != 0.0 ? sampling[m] : 1.0);
+            averaged_intensity.emplace(q_mesh(m), average);
+        }
+        return averaged_intensity;
+    }
+
+};
+
+
+/**
+ * A policy for collecting samples. To be used together with StructureFactor class templates.
+ *
+ * @tparam T float or double
+ */
+template <typename T> class SamplingPolicy {
+  public:
+    struct sampled_value {
+        T value;
+        T weight;
+    };
+    typedef std::map<T, sampled_value> TSampledValues;
```

Comment:
"TSampledValues" does to reflect that it's a map containing both keys and values.

Reply:
Any suggestion on the name? (This is the last point to be resolved.)

Reply:
TSampledValueMap? Analogous we have TParticleVector in Space.

Reply:
Done.